### DISCUSSION TITLE
Asyncronous loops to perform SSE

### DISCUSSION THREAD

**Discussion by fonty422** [fastapi/fastapi#7688] [Questions] [answered]
I'm refactoring from Flask with flask-socketIO and threads to use fastapi with asyncio. The basics are excellent, but I have the need for server side event initiation. I'm trying to avoid websockets and want to make it purely http.
A good example is a simple heartbeat signal as a 1/0 that is sent each second to all connected. I can't seem to figure out how to make that happen. There are a few loops like this in my old code that are purely server driven events.
Is this kind of thing possible without the need for websockets?

**Comment by transfluxus** [comment]
I am working on something similar.
My approach now is after accepting the connection I am storing the websocket object in an dict (user_id) as key. By this I can grab them from any other point in my app and send messages

**Comment by fonty422** [comment]
Yep... although I'm trying not to use websockets just pure server-sent-events.

**Comment by fonty422** [comment]
I managed to get nice behavior using Flask with the following:
```
#flask app
import flask, time
from flask import Response, render_template

def get_message():
  '''this could be any function that blocks until data is ready'''
  
  time.sleep(1)
  s = time.ctime(time.time())
  return s

@app.route('/stream', methods=["GET"])
def stream():
  def eventStream():
    while True:
      # wait for source data to be available, then push it
      yield 'data: {}\n\n'.format(get_message())
  return Response(eventStream(), mimetype="text/event-stream")
```

and the following on the client side (React)
```
const timestamp = new EventSource("http://localhost/stream", {method: "get", mode: 'cors'})
timestamp.onmessage = (data) => {
  console.log(data.data)
}
```

But I can't implement the same with FastAPI. It does the whole loop then sends it all in one chunk rather than in 1s intervals.

But I'd really rather use FastAPI to achieve the same

**Comment by euri10** [comment, marked as answer]
Have you looked if https://github.com/sysid/sse-starlette would help?

**Comment by fonty422** [comment]
Oooo... I had stupidly thought it was for websockets, but I'm happily mistaken! Thanks @euri10

**Comment by transfluxus** [comment]
what are pure server-sent-events? If you think about HTTP2.0 pushes, that is not the right thing here.
websockets are definitely a working solution here, and you can do repeated tasks with FastAPI utils
https://fastapi-utils.davidmontague.xyz/user-guide/repeated-tasks/
(I myself haven't that feature tho)

**Comment by fonty422** [comment]
@transfluxus, what a nice solution. It will work well now that I've also been pointed to the inbuilt websocket option. I was wanting to avoid websockets to avoid using another package, but this looks neat and simple as an always on timed event generator.

**Comment by fonty422** [comment]
Both are good solutions, so I can close this now.

**Comment by tiangolo** [comment]
Thanks for the help here @euri10 and @transfluxus ! :cake: :bowing_man: 

@fonty422 I suspect the equivalent to what you were using would be a `StreamingResponse`, maybe you can check that too: https://fastapi.tiangolo.com/advanced/custom-response/#streamingresponse

Anyway, thanks for reporting back and closing the issue :+1:

**Comment by Kludex** [comment]
`sse-starlette` handles some issues when reloading/closing the application.
